// angular-feather/esm2015/lib/feather.component.js
import { Component, ElementRef, Input, Inject, ChangeDetectorRef, SecurityContext, } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { Icons } from './icons.provider';
import { uppercamelcase } from './utils';
import * as i0 from "@angular/core";
import * as i1 from "./icons.provider";
import * as i2 from "@angular/platform-browser";
export class FeatherComponent {
    constructor(elem, changeDetector, icons, sanitizer) {
        this.elem = elem;
        this.changeDetector = changeDetector;
        this.icons = icons;
        this.sanitizer = sanitizer;
    }
    ngOnChanges(changes) {
        // icons are provided as an array of objects because of "multi: true"
        const icons = Object.assign({}, ...this.icons);
        const svg = icons[uppercamelcase(changes.name.currentValue)] || '';
        if (!svg) {
            console.warn(`Icon not found: ${changes.name.currentValue}\n` +
                `Refer to documentation on https://github.com/michaelbazos/angular-feather`);
        }
        // Since the icons are precompiled we can trust them as safe html.
        this.elem.nativeElement.innerHTML = this.sanitizer.sanitize(SecurityContext.HTML, this.sanitizer.bypassSecurityTrustHtml(svg));
        this.changeDetector.markForCheck();
    }
}
FeatherComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: FeatherComponent, deps: [{ token: ElementRef }, { token: ChangeDetectorRef }, { token: Icons }, { token: DomSanitizer }], target: i0.ɵɵFactoryTarget.Component });
FeatherComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.2.17", type: FeatherComponent, selector: "i-feather, feather-icon", inputs: { name: "name" }, usesOnChanges: true, ngImport: i0, template: "<ng-content></ng-content>\n", styles: [":host{display:inline-block;width:24px;height:24px;fill:none;stroke:currentColor;stroke-width:2px;stroke-linecap:round;stroke-linejoin:round}\n"] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: FeatherComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'i-feather, feather-icon',
                    templateUrl: './feather.component.html',
                    styleUrls: ['./feather.component.scss'],
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i0.ChangeDetectorRef, decorators: [{
                    type: Inject,
                    args: [ChangeDetectorRef]
                }] }, { type: i1.Icons, decorators: [{
                    type: Inject,
                    args: [Icons]
                }] }, { type: i2.DomSanitizer, decorators: [{
                    type: Inject,
                    args: [DomSanitizer]
                }] }]; }, propDecorators: { name: [{
                type: Input
            }] } });
